### PR TITLE
Added timeout to blackbox exporter

### PR DESCRIPTION
**Pull request by Conorbro**
Added timeout set by Prometheus header to blackbox exporter.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -108,26 +107,38 @@ func init() {
...
 	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
+	defer cancel()
+	r = r.WithContext(ctx)
+	probeHandler(w, r, c)
+}
 
-	log.Infoln("Starting blackbox_exporter", version.Info())
-	log.Infoln("Build context", version.BuildContext())
+var timeoutOffset = flag.Float64("timeout-offset", 0.5, "Offset to subtract from timeout in seconds.")
```
Comment (brian-brazil) [inline main.go]: All the flags should be together.

File: config_test.go [hunk 2]
```diff
@@ -25,3 +29,23 @@ func TestLoadBadConfig(t *testing.T) {
 		t.Errorf("\nexpected:\n%v\ngot:\n%v", expected, err.Error())
 	}
 }
+
+func TestPrometheusTimeout(t *testing.T) {
+	initConfig()
```
Comment (brian-brazil) [inline config_test.go]: This is pulling the file off the filesystem, it'd be better in the test to pass in the config to be used to probeRequestHandler. It's already setup to allow you to inject a config.

File: config_test.go [hunk 3]
```diff
@@ -25,3 +29,23 @@ func TestLoadBadConfig(t *testing.T) {
...
+
+	req, err := http.NewRequest("GET", "/probe?module=http_2xx&target=http%3A%2F%2Fprometheus.io", nil)
```
Comment (brian-brazil) [inline config_test.go]: We shouldn't presume internet access, see the http tests for how to setup a test server

File: main.go [hunk 4]
```diff
@@ -52,6 +54,10 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
 		return err
 	}
 
+	if err := checkOverflow(c.XXX, "config"); err != nil {
```
Comment (brian-brazil) [inline main.go]: This belongs in Unmarshal

File: config.go [hunk 5]
```diff
@@ -98,6 +99,18 @@ type DNSRRValidator struct {
 	XXX map[string]interface{} `yaml:",inline"`
 }
 
+func getModule(c *Config, r *http.Request, w http.ResponseWriter) Module {
```
Comment (brian-brazil) [inline config.go]: This is not related to parsing the config, it doesn't belong in this file

File: config_test.go [hunk 6]
```diff
@@ -25,3 +29,23 @@ func TestLoadBadConfig(t *testing.T) {
 		t.Errorf("\nexpected:\n%v\ngot:\n%v", expected, err.Error())
 	}
 }
+
+func TestPrometheusTimeout(t *testing.T) {
```
Comment (brian-brazil) [inline config_test.go]: This doesn't belong in this file

File: config_test.go [hunk 7]
```diff
@@ -25,3 +29,23 @@ func TestLoadBadConfig(t *testing.T) {
...
+		t.Fatal(err)
+	}
+
+	req.Header.Set("X-Prometheus-Scrape-Timeout-Seconds", "5")
+
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(probeRequestHandler)
+	handler.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusOK {
```
Comment (brian-brazil) [inline config_test.go]: Make a endpoint that's purposefully slow and see it timeout

File: main.go [hunk 8]
```diff
@@ -76,15 +82,8 @@ func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
...
-	moduleName := params.Get("module")
-	if moduleName == "" {
-		moduleName = "http_2xx"
-	}
-	module, ok := conf.Modules[moduleName]
-	if !ok {
-		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
-		return
-	}
+	module := getModule(conf, r, w)
```
Comment (brian-brazil) [inline main.go]: I don't think factoring this out is gaining much

File: main.go [hunk 9]
```diff
@@ -108,26 +107,38 @@ func init() {
 	prometheus.MustRegister(version.NewCollector("blackbox_exporter"))
 }
 
-func main() {
+func probeRequestHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (brian-brazil) [inline main.go]: I'd leave this inline with no changes. Splitting the logic across two functions is confusing.

File: main.go [hunk 10]
```diff
@@ -108,26 +107,38 @@ func init() {
...
-		fmt.Fprintln(os.Stdout, version.Print("blackbox_exporter"))
-		os.Exit(0)
+	module := getModule(c, r, w)
+
+	if module.Timeout.Seconds() < timeoutSeconds {
+		timeoutSeconds = module.Timeout.Seconds()
 	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
+	defer cancel()
+	r = r.WithContext(ctx)
```
Comment (brian-brazil) [inline main.go]: You need to use this context down in the probes

**Comment by Conorbro** [comment]
@brian-brazil

File: dns_test.go [hunk 11]
```diff
@@ -118,7 +119,10 @@ func TestRecursiveDNSResponse(t *testing.T) {
 			test.Probe.TransportProtocol = protocol
 			registry := prometheus.NewPedanticRegistry()
 			registry.Gather()
-			result := probeDNS(addr.String(), Module{Timeout: time.Second, DNS: test.Probe}, registry)
+
+			testCTX, cancel := context.WithTimeout(context.Background(), time.Duration(10)*time.Second)
```
Comment (brian-brazil) [inline dns_test.go]: you can do `10 * time.Second`

File: main.go [hunk 12]
```diff
@@ -76,15 +78,15 @@ func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
 		return
 	}
 
-	moduleName := params.Get("module")
+	moduleName := r.URL.Query().Get("module")
```
Comment (brian-brazil) [inline main.go]: this is available as params from above

File: main.go [hunk 13]
```diff
@@ -108,15 +112,55 @@ func init() {
 	prometheus.MustRegister(version.NewCollector("blackbox_exporter"))
 }
 
+func probeRequestHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (brian-brazil) [inline main.go]: You're still arbitrarily splitting logic across probeRequestHandler and probeHandler. All this logic should be in one function, with a small shim anonymous function in the HandleFunc.

File: main.go [hunk 14]
```diff
@@ -108,15 +112,55 @@ func init() {
...
+	moduleName := r.URL.Query().Get("module")
+	if moduleName == "" {
+		moduleName = "http_2xx"
+	}
+	module, ok := c.Modules[moduleName]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
+	}
+
+	if module.Timeout.Seconds() < timeoutSeconds && module.Timeout.Seconds() > 0 {
```
Comment (brian-brazil) [inline main.go]: Keep the timeout logic together

File: main.go [hunk 15]
```diff
@@ -108,15 +112,55 @@ func init() {
...
+		timeoutSeconds = module.Timeout.Seconds()
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
+	defer cancel()
+	r = r.WithContext(ctx)
+
+	probeHandler(ctx, w, r, c)
+}
+
+var timeoutOffset = flag.Float64("timeout-offset", 0.5, "Offset to subtract from timeout in seconds.")
```
Comment (brian-brazil) [inline main.go]: Keep all the flags together, they should be up near the top of this file in a var section.

File: main_test.go [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+		"http_2xx": Module{
+			Prober:  "http",
+			Timeout: time.Duration(10 * time.Second),
+		},
+	},
+}
+
+func TestPrometheusTimeout(t *testing.T) {
+
+	sc.C = c
```
Comment (brian-brazil) [inline main_test.go]: Changing globals in tests is best avoided where possible. Pass in what you need to the function.

File: main_test.go [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	req.Header.Set("X-Prometheus-Scrape-Timeout-Seconds", "1")
+
+	rr := httptest.NewRecorder()
+	handler := http.HandlerFunc(probeRequestHandler)
+	handler.ServeHTTP(rr, req)
+
+	if status := rr.Code; status != http.StatusInternalServerError {
```
Comment (brian-brazil) [inline main_test.go]: This should return a 200.

**Comment by Conorbro** [comment]
@brian-brazil

File: main.go [hunk 18]
```diff
@@ -176,6 +199,10 @@ func main() {
 				http.Error(w, fmt.Sprintf("failed to reload config: %s", err), http.StatusInternalServerError)
 			}
 		})
+	http.Handle("/metrics", prometheus.Handler())
+	http.HandleFunc("/probe", func(w http.ResponseWriter, r *http.Request) {
+		probeHandler(w, r, nil)
```
Comment (brian-brazil) [inline main.go]: The config should continue to be passed in here

File: main.go [hunk 19]
```diff
@@ -60,7 +72,44 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
...
+
+	// If a timeout is configured via the Prometheus header, add it to the request.
+	var prometheusTimeout string
+	if r.Header["X-Prometheus-Scrape-Timeout-Seconds"] != nil {
+		prometheusTimeout = r.Header["X-Prometheus-Scrape-Timeout-Seconds"][0]
+	}
+
+	timeoutSeconds, err := strconv.ParseFloat(prometheusTimeout, 64)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to parse timeout from Prometheus header: %s", err), http.StatusInternalServerError)
```
Comment (brian-brazil) [inline main.go]: if you're using http.Error you need to `return` right after it.

File: main.go [hunk 20]
```diff
@@ -60,7 +72,44 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
...
+	}
+	if timeoutSeconds == 0 {
+		timeoutSeconds = 10
+	}
+
+	if module.Timeout.Seconds() < timeoutSeconds && module.Timeout.Seconds() > 0 {
+		timeoutSeconds = module.Timeout.Seconds()
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutSeconds-*timeoutOffset)*time.Second)
```
Comment (brian-brazil) [inline main.go]: You're using precision here as time.Duration is an int64. Multiplying it by 1e9 before casting would keep the precision.

File: main.go [hunk 21]
```diff
@@ -76,17 +125,9 @@ func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
...
-		moduleName = "http_2xx"
-	}
-	module, ok := conf.Modules[moduleName]
-	if !ok {
-		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
-		return
-	}
 	prober, ok := Probers[module.Prober]
 	if !ok {
+		fmt.Println("JHere")
```
Comment (brian-brazil) [inline main.go]: test code

File: tcp.go [hunk 22]
```diff
@@ -48,23 +48,26 @@ func dialTCP(target string, module Module, registry *prometheus.Registry) (net.C
...
 	if !module.TCP.TLS {
-		return dialer.Dial(dialProtocol, dialTarget)
+		return dialer.DialContext(ctx, dialProtocol, dialTarget)
 	}
 	tlsConfig, err := config.NewTLSConfig(&module.TCP.TLSConfig)
 	if err != nil {
 		return nil, err
 	}
+	timeoutDeadline, _ := ctx.Deadline()
+	dialer.Timeout = timeoutDeadline.Sub(time.Now())
```
Comment (brian-brazil) [inline tcp.go]: You can use dialer.Deadline to make it simpler

File: tcp.go [hunk 23]
```diff
@@ -48,23 +48,26 @@ func dialTCP(target string, module Module, registry *prometheus.Registry) (net.C
...
 	return tls.DialWithDialer(dialer, dialProtocol, dialTarget, tlsConfig)
 }
 
-func probeTCP(target string, module Module, registry *prometheus.Registry) bool {
+func probeTCP(ctx context.Context, target string, module Module, registry *prometheus.Registry) bool {
 	probeSSLEarliestCertExpiry := prometheus.NewGauge(prometheus.GaugeOpts{
 		Name: "probe_ssl_earliest_cert_expiry",
 		Help: "Returns earliest SSL cert expiry date",
 	})
 	registry.MustRegister(probeSSLEarliestCertExpiry)
```
Comment (brian-brazil) [inline tcp.go]: Hmm, this should only be registered if we set it

**Comment by Conorbro** [comment]
@brian-brazil

File: main.go [hunk 24]
```diff
@@ -60,7 +72,45 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
 	return nil
 }
 
-func probeHandler(w http.ResponseWriter, r *http.Request, conf *Config) {
+func probeHandler(w http.ResponseWriter, r *http.Request, c *Config) {
+
+	if c == nil {
```
Comment (brian-brazil) [inline main.go]: You don't need this.

File: main.go [hunk 25]
```diff
@@ -60,7 +72,45 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
...
+		sc.RUnlock()
+	}
+
+	moduleName := r.URL.Query().Get("module")
+	if moduleName == "" {
+		moduleName = "http_2xx"
+	}
+	module, ok := c.Modules[moduleName]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Unknown module %q", moduleName), 400)
```
Comment (brian-brazil) [inline main.go]: Return after sending a http error

File: main.go [hunk 26]
```diff
@@ -60,7 +72,45 @@ func (sc *SafeConfig) reloadConfig(confFile string) (err error) {
...
+		return
+	}
+	if timeoutSeconds == 0 {
+		timeoutSeconds = 10
+	}
+
+	if module.Timeout.Seconds() < timeoutSeconds && module.Timeout.Seconds() > 0 {
+		timeoutSeconds = module.Timeout.Seconds()
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration((timeoutSeconds-*timeoutOffset)*1e9)*time.Second)
```
Comment (brian-brazil) [inline main.go]: Don't multiply by time.Second, the 1e9 already covers that

File: main.go [hunk 27]
```diff
@@ -176,6 +199,10 @@ func main() {
 				http.Error(w, fmt.Sprintf("failed to reload config: %s", err), http.StatusInternalServerError)
 			}
 		})
+	http.Handle("/metrics", prometheus.Handler())
+	http.HandleFunc("/probe", func(w http.ResponseWriter, r *http.Request) {
+		probeHandler(w, r, sc.C)
```
Comment (brian-brazil) [inline main.go]: You need to protect this access to sc.C with the mutex

**Comment by brian-brazil** [comment]
This will need to be documented in the README too

**Comment by Conorbro** [comment]
@brian-brazil

File: README.md [hunk 28]
```diff
@@ -39,6 +39,8 @@ Additionally, an [example configuration](https://github.com/prometheus/blackbox_
 HTTP, HTTPS (via the `http` prober), DNS, TCP socket and ICMP (see permissions section) are currently supported.
 Additional modules can be defined to meet your needs.
 
+The timeout of each probe may be configured using `timeout` in the Blackbox exporter config file or preferably by setting the `scrape_timeout` in the [Prometheus config](https://prometheus.io/docs/operating/configuration/#configuration-file) itself.
```
Comment (brian-brazil) [inline README.md]: This doesn't quite capture the logic. 

The timeout of each probe is automatically determined from the `scrape_timeout` in the Prometheus config, slightly reduced to allow for network delays. This can be further limited by the `timeout` in the Blackbox exporter config file. If neither is specified, it defaults to 10 seconds.

File: main.go [hunk 29]
```diff
@@ -176,6 +194,12 @@ func main() {
 				http.Error(w, fmt.Sprintf("failed to reload config: %s", err), http.StatusInternalServerError)
 			}
 		})
+	http.Handle("/metrics", prometheus.Handler())
+	http.HandleFunc("/probe", func(w http.ResponseWriter, r *http.Request) {
+		sc.RLock()
```
Comment (brian-brazil) [inline main.go]: The mutex is taken for too long here, this will block reloads while a probe is in progress. Only hold the mutex for enough time to copy over C.

**Comment by Conorbro** [comment]
@brian-brazil

File: main.go (hunk 27 above)
Comment (brian-brazil) [inline main.go]: This comment remains

**Comment by brian-brazil** [comment]
The CI doesn't like this, can you run make format?

**Comment by Conorbro** [comment]
@brian-brazil

**Comment by brian-brazil** [comment]
Thanks!
